### PR TITLE
Virtual authenticators extension support

### LINKED ISSUE #1471: Virtual authenticator support for extensions
At the moment, the [User Agent Automation section](https://w3c.github.io/webauthn/#sctn-automation) does not say anything about treatment of extensions by virtual authenticators. As user agents implement the extensions, we should strive to support them through the virtual authenticators so that relying parties can write automated tests and we can ensure interoperability with web platform tests.

My proposal is to add
- for every defined extension, a capability with the name `webauthn:$extension_identifier` signalling virtual authenticator support for the extension, and
- an `extensions` array to the `Authenticator Configuration` object, indicating which extensions must be supported by the specific virtual authenticator. Extensions not present in the array should not be supported (this is to test that a given relying party or user agent does the right thing when the underlying authenticator does not support an extension).
- when defining an extension, the text
  > User agents implementing $extension SHOULD support the `webauthn:$extension_identifier` webdriver capability

### PR DESCRIPTION
**Pull request by nsatragno**
Add support for authenticator extensions on WebDriver virtual authenticators.

This comes in the form of:
* A capability for each extension signalling support.
* An option to toggle which extensions are supported by a given virtual authenticator.

Fixes #1471

***
<a href="https://pr-preview.s3.amazonaws.com/nsatragno/webauthn/pull/1472.html" title="Last updated on Sep 2, 2020, 3:52 PM UTC (ea9ad16)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/1472/ad8d1d0...nsatragno:ea9ad16.html" title="Last updated on Sep 2, 2020, 3:52 PM UTC (ea9ad16)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -5943,6 +6012,11 @@ Each stored [=virtual authenticator=] has the following properties:
     [=User Verification=] will always succeed. If set to [FALSE], it will fail.
 
     Note: This property has no effect if |hasUserVerification| is set to [FALSE].
+: |extensions|
+:: A string array containing the [=extension identifiers=] supported by the [=Virtual Authenticator=].
+
+[=Virtual authenticators=] MUST support all [=authenticator extensions=] present in its |extensions| array.
+They MUST NOT support any [=authenticator extension=] not present in its |extensions| array.
```
Comment (equalsJeffH) [inline index.bs:6019]: If the virtual authnr does not support any extensions, is this property's value an empty array, or is the property absent? Ah, I see below in "Add Virtual Authenticator" that the default is an empty array.

Comment (emlun) [inline index.bs:6019]: There's a mix of plural and singular here. I would suggest:

```suggestion
A [=virtual authenticators=] MUST support all [=authenticator extensions=] present in its |extensions| array.
It MUST NOT support any [=authenticator extension=] not present in its |extensions| array.
```

File: index.bs [hunk 2]
```diff
@@ -5916,6 +5917,74 @@ When [=matching capabilities=], the extension-specific steps to match `"webauthn
...
+                <th>Capability</th>
+                <th>Key</th>
+                <th>Value Type</th>
+                <th>Description</th>
+            </tr>
+        </thead>
+        <tbody>
+            <tr>
+                <td>FIDO AppID Extension Support</td>
+                <td>`"webauthn:appid"`</td>
```
Comment (emlun) [inline index.bs:5937]: How about using a second namespace level for extensions, like `webauthn:extension:appid` or `webauthn:ext:appid`? I'm thinking capabilities like `webauthn:pin-userVerification` or something like that might be useful at some point, maybe for features added in the future, and it seems appropriate to separate the extensions from other features.

File: index.bs (hunk 1 above)
Comment (emlun) [reply index.bs:6019]: Might also be appropriate to indent this as a second paragraph of the `|extensions|` definition? I'm not sure.

**Review by emlun — COMMENTED** [review]
Maybe this should also include a new step in the remote end steps for "add credential" and "get credentials"?

File: index.bs [hunk 3]
```diff
@@ -5916,6 +5917,56 @@ When [=matching capabilities=], the extension-specific steps to match `"webauthn
         the match is unsuccessful.
     2. Otherwise, the match is successful.
 
+### <dfn>Authenticator Extension Capabilities</dfn> ### {#sctn-authenticator-extension-capabilities}
+
+Additionally, [=extension capabilities=] are defined for every [=authenticator extension=] defined in this specification:
```
Comment (emlun) [inline index.bs:5922]: This "every" is no longer accurate after the latest update. :)

File: index.bs [hunk 4]
```diff
@@ -5943,6 +5994,15 @@ Each stored [=virtual authenticator=] has the following properties:
...
     Note: This property has no effect if |hasUserVerification| is set to [FALSE].
+: |extensions|
+:: A string array containing the [=extension identifiers=] supported by the [=Virtual Authenticator=].
+
+    A [=Virtual authenticator=] MUST support all [=authenticator extensions=] present in its |extensions| array.
+    It MUST NOT support any [=authenticator extension=] not present in its |extensions| array.
+: |uvm|
+:: A {{UvmEntries}} array to be set as the [=authenticator extension output=] when processing the [=User Verification Method=] extension.
+
+   Note: This property has no effect if the [=Virtual Authenticator=] does not support the [=User Verification Method=] extension.
```
Comment (emlun) [inline index.bs:6005]: Heh, this needs one more space of indentation to render as intended.

```suggestion
    Note: This property has no effect if the [=Virtual Authenticator=] does not support the [=User Verification Method=] extension.
```

Comment (nsatragno) [reply index.bs]: Something something meaningful indentation is evil. Done!

File: index.bs (hunk 3 above)
Comment (nsatragno) [reply index.bs]: I think it's still true.

> An extension can also be an *authenticator extension*, meaning that the extension involves communication with and processing by the authenticator.

The extensions I left here are precisely those that involve processing by the authenticator, i.e. those for which "Authenticator extension processing" is not empty.

--

I agree this text can be confusing so I added a clarification.

Comment (emlun) [reply index.bs]: Oh right, sorry, this was indeed accurate. So I don't think the clarification is really needed, but I don't mind keeping it in either.

**Review by emlun — APPROVED** [review]
LGTM, thanks @nsatragno!

**Comment by jcjones** [comment]
So I just want to make sure I understand properly - the idea here is that _if_ the WPTs can expect the Virtual Authenticator to support a given extension, _then_ the capabilities reported via WebDriver to the test should indicate that extension?

**Comment by nsatragno** [comment]
> So I just want to make sure I understand properly - the idea here is that if the WPTs can expect the Virtual Authenticator to support a given extension, then the capabilities reported via WebDriver to the test should indicate that extension?

Yes, that is correct.

**Review by jcjones — APPROVED** [review]

**Review by akshayku — APPROVED** [review]

**Pull request merged**